mercury_engine_data_structures/formats/bcmdl.py

```diff
@@ -0,0 +1,593 @@
+import construct
+from construct import Construct
+from construct.core import (
+    Array, Byte, Bytes, Const, Construct, Container, Flag, GreedyBytes, Hex, If, IfThenElse,
+    Int16ul, Int32sl, Int32ul, Int64ul, ListContainer, Pointer, Peek, RepeatUntil, stream_seek, Struct, Tell
+)
+
+from mercury_engine_data_structures.common_types import Float, StrId, make_vector
+from mercury_engine_data_structures.formats import BaseResource
+from mercury_engine_data_structures.game_check import Game
+
+# Partial implementation to generate material variants
+# Based off Joschuka's MMDL implementation in Noesis
+# https://github.com/Joschuka/fmt_mmdl/blob/main/fmt_mmdl.py
+
+# prints a pointer's hex value in containers, to make finding it in a binary easier
+Ptr = Hex(Int64ul)
+
+# pad 0xFF bytes to an offset of 8
+PadTo8B = Struct(
+    _cur_pos = Tell,
+    padding = construct.Padding((8 - construct.this._cur_pos) % 8, pattern=b'\xff')
+)
+
+# an entry in a linked-list typically pointed to by TOC.
+# next entry should be whats pointed to but in practice all the entries are contiguous
+# so we just need to break when next==0
+TOC_subentry = Struct(
+    ptr = Ptr,
+    next = Ptr
+)
+
+# info on a section of the decompressed vertex buffer
+# example, semantic=0, offset=0x48, count=7 means starting from buffer+0x48, read 7 XYZ positions
+# a little complex in context, see below
+# https://github.com/Joschuka/fmt_mmdl/blob/c8c2ff9f10d95c6d4eedbbe45abea9dd535cf22e/fmt_mmdl.py#L695
+Vertex_Info_entry = Struct(
+    semantic = Int32ul, # 0-8, controls different things (0=vert positions, 1=vert normals, etc)
+    offset = Int32ul, # the offset from the start of decompressed Vert_Buffer
+    data_type = Int16ul, # always 3?
+    count = Int16ul, # number of entries
+    unk = Int32ul
+)
+
+# vertex buffer
+Vert_Buffer = Struct(
+    _gzip_header = Peek(Int32ul),
+    buf = IfThenElse(
+        construct.this._gzip_header != 559903, # gzip header
+        Bytes(construct.this._.buffer_size),
+        Bytes(construct.this._.comp_size) # TODO inflate this and have it correctly deflate
+    ),
+    _padding = PadTo8B
```

Comment:
You can add an `construct.Aligned` around the entire `Struct`, it should do this padding for you